### PR TITLE
ScanFile expression and visitor for CDF

### PR DESCRIPTION
**Pull request by OussamaSaoudi-db**
## What changes are proposed in this pull request?
This PR introduces four concepts:
- `cdf_scan_row_schema`: This is the schema that engine data will be transformed into at the end of the log replay phase. This schema prunes the log schema down only to the fields necessary to produce CDF columns. 
- `cdf_scan_row_expression`: This is a function that generates an expression to transform an engine data into the `cdf_scan_row_schema`. The function takes timestamp and commit number as arguments because it inserts these as columns into the output engine data.
- `CDFScanFile`: This is a type that holds all the information needed to read a data file and generate its CDF rows. It holds path, deletion vector, the type of action, and the paired remove deletion vector. The action type is encoded as an enum `CDFScanFileType`
- `CDFScanFileVisitor`: This is a visitor that reads engine data with the `cdf_scan_row_schema` and constructs `CDFScanFile`s.

This PR is only for internal use, and is only expected to be used by `TableChangesScan::execute` when it is implemented. Engines must *not* use the visitor nor `CDFScanFile`.

### This PR affects the following public APIs
No public APIs are changed. 

## How was this change tested?
I generate a table with add, remove and cdc actions. Then:
- The table is read,
- The engine data is transformed using `table_changes_action_iter` which in transforms the engine data into the `cdf_scan_row_schema` using the `cdf_scan_row_expression`
- The transformed engine data is read again using the `CDFScanFileVisitor` and assert that the `CDFScanFile`s are as expected.

This test checks the following cases:
- A remove with `None` partition values. An empty hashmap for partition values should be used.
- A remove with partition values.
- An add/remove DV pair. This should place the correct remove dv into the add's CdfScanFile.
- The visitor extracts the correct timestamp and commit version for each file.

### CODE REVIEW

File: kernel/src/table_changes/scan_file.rs
Comment (nicklan) [inline kernel/src/table_changes/scan_file.rs:56]: i prefer to call this `scan_type` or similar

File: kernel/src/table_changes/scan_file.rs
Comment (nicklan) [inline kernel/src/table_changes/scan_file.rs:54]: Maybe rename this `CDFScanFile` or something. I think the types are going to get confusing if they have the same name but exist in different modules

File: kernel/src/table_changes/scan_file.rs
Comment (nicklan) [inline kernel/src/table_changes/scan_file.rs:246]: probably doesn't need to be `pub(crate)`. I think it should only be used in this file

**Review by nicklan — COMMENTED** [review]
This looks fine. Please note in the PR that this is _internal_ for now. That is, at some point we may want to expose this to engines, but for the first draft of CDF we are _only_ allowing a full `execute()` which just uses this internally and doesn't expose it

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:246]: This will be used by the `LogReplayScanner` since this is the output schema for the transformed EngineData.

**Comment by OussamaSaoudi-db** [comment]
todo: rename to `scan_data_*` instead of `scan_row_*`

**Comment by OussamaSaoudi-db** [comment]
TODO: Remove the comment on schema because that can become out of sync

**Comment by OussamaSaoudi-db** [comment]
TODO: Make these changes: https://github.com/delta-io/delta-kernel-rs/pull/540#discussion_r1871482361

**Review by scovich — COMMENTED** [review]
quick drive-by, but will wait for pending changes to do a full review

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:26]: I was considering introducing two types:
```
#[allow(unused)]
pub(crate) struct UnresolvedCDFScanFile {
    scan_file: CDFScanFile,
    remove_dvs: Arc<HashMap<String, DvInfo>>,
}
#[allow(unused)]
pub(crate) struct ResolvedCDFScanFile {
    scan_file: CDFScanFile,
    selection_vector: Vec<bool>,
}
```

Because the step after this will be to resolve deletion vectors. Is this sensible naming?

**Comment by OussamaSaoudi-db** [comment]
@scovich  you had a question about whether we need to use [unsigned ints to track commit versions](https://github.com/delta-io/delta-kernel-rs/pull/540#discussion_r1871460132). I'll check the spec at some point.

Visitor code expects that all numbers are signed ints, but we declare `type Version = u64`. I'd been keeping versions as `u64`, but I think we should come up with a consistent approach.

**Comment by scovich** [comment]
> @scovich you had a question about whether we need to use [unsigned ints to track commit versions](https://github.com/delta-io/delta-kernel-rs/pull/540#discussion_r1871460132). I'll check the spec at some point.
> 
> Visitor code expects that all numbers are signed ints, but we declare `type Version = u64`. I'd been keeping versions as `u64`, but I think we should come up with a consistent approach.

Can we make a tracking issue, with the goal to decide once and for all our stance on signed vs. unsigned?
Or maybe https://github.com/delta-io/delta-kernel-rs/issues/174 _is_ that issue and we just need to do something about it.

File: kernel/src/table_changes/scan_file.rs
Comment (nicklan) [inline kernel/src/table_changes/scan_file.rs:191]: it's a really nice way to handle this :)

**Review by nicklan — APPROVED** [review]
given that this is all internal, stamping for now and we can iterate later if needed

**Review by scovich — APPROVED** [review]
Approach looks sound. Several nits to address (especially the one about schema prining/nullability), but it should be quick and easy.

**Review by zachschuermann — APPROVED** [review]
Can we add a `scan_data_to_scan_file` test (just something through that API) otherwise LGTM!

**Comment by zachschuermann** [comment]
(and we can take some of the todo's above from my other reviews as follow-up or just quickly in this PR, either way)

**Comment by zachschuermann** [comment]
and this shouldn't be a breaking change right?

File: kernel/src/table_changes/scan_file.rs
Comment (scovich) [inline kernel/src/table_changes/scan_file.rs:72]: I don't understand why this method can't take `scan_data` directly (instead of breaking it up and passing (almost) all its fields as individual args? Then we could directly probe the remove_dvs while visiting adds, and `CDFScanFile` could hold dv_info for both add and remove, and the next PR would have a much easier time resolving DVs (no need to pass the hashmap to it).

Is it attempting to preserve similarity to the analagous `visit_scan_files` or something?

**Review by scovich — COMMENTED** [review]
Late find while reviewing https://github.com/delta-io/delta-kernel-rs/pull/568

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:87]: Good point, I was preserving the similarity with `visit_scan-files`. iirc this part of the code has ffi-related reasons for being the way that it is 🤔  I'll look into it, but I'll ping @nicklan if he has input

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs]: Added docs

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: > should we just use DeletionVectorDescriptor::to_schema instead

I'd like to, but we hard code the number of fields in the visitor getters. So any change to the DV schema would break the visitor.

The schema stuff makes sense! I'll keep the top level fields nullable and patch up the nested ones.

Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: Also given that partitionValues is technically nullable for removes, I'm constructing a default map if I don't find anything. Is this necessary/sensible?
```rust
let partition_values = partition_values.unwrap_or_else(Default::default);
```

Comment (scovich) [reply kernel/src/table_changes/scan_file.rs:220]: > Found unmasked nulls for non-nullable StructArray field \"path\"

That error indicates that `add` was non-null but `path` was null, which is definitely not a good thing. Do you have a backtrace? I wonder if this tickled another latent bug in our arrow data code, similar to the one that @nicklan hit a while back?

Comment (scovich) [reply kernel/src/table_changes/scan_file.rs:220]: (also just in case -- did you verify that the underlying data are actually correct? we have tripped over invalid data a couple times in the past)

Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: Hmm I'll try to track this down and report back 🫡

Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: Okay I've narrowed down the point in the code where we fail for this. It's actually in the expression transformation, and not the visitor!

The specific line is [here](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/engine/arrow_expression.rs#L411). I don't see anything that handles nested nullability, but maybe arrow is supposed to do that on its own?

I think I've got a reasonably MRE. I'll post an issue.

Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: I submitted an issue. Would love for you to sanity check it @scovich: 
https://github.com/delta-io/delta-kernel-rs/issues/578

Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:220]: I'll punt on this for now and keep everything nullable. This is what we do in the normal read path, so I think it should be okay for now.

**Comment by OussamaSaoudi-db** [comment]
@zachschuermann @scovich  I updated the test to use the log replay. PTAL

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs]: I ended up switching it and it simplified it nicely. Ig we can deal with engine support later given that we aren't targeting that rn

File: kernel/src/table_changes/scan_file.rs
Comment (scovich) [reply kernel/src/table_changes/scan_file.rs:189]: attn @nick

**Review by scovich — APPROVED** [review]

File: kernel/src/table_changes/scan_file.rs
Comment (zachschuermann) [inline kernel/src/table_changes/scan_file.rs:117]: why ref arc? should this just be arc?

File: kernel/src/table_changes/scan_file.rs
Comment (zachschuermann) [inline kernel/src/table_changes/scan_file.rs:221]:
```suggestion
/// [`cdf_scan_row_schema`]. This is the expression used to create [`TableChangesScanData`].
```

**Review by zachschuermann — APPROVED** [review]
LGTM couple questions/nits!

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:21]: Yeah I think I'll leave this to the final `execute`  pr.

File: kernel/src/table_changes/scan_file.rs
Comment (OussamaSaoudi-db) [reply kernel/src/table_changes/scan_file.rs:117]: Didn't see a reason to do an entire arc clone in the loop. I changed this to instead have `&'a HashMap<String, DvInfo>`, and we use the arc `as_ref`

**Pull request merged**
